zerver/lib/bot_lib.py

```diff
@@ -32,6 +34,58 @@ def get_bot_handler(service_name):
     bot_module = importlib.import_module(bot_module_name)  # type: Any
     return bot_module.handler_class()
 
+class StateHandler(object):
+    # Cache the state size of each bot so we don't have to do expensive accumulations
+    # each time the bot receives a message.
+    state_sizes = {}  # type: Dict[int, int]
+    state_size_limit = 10000000   # type: int # TODO: Store this in the server configuration model.
+
+    def __init__(self, user_profile):
+        # type: (UserProfile) -> None
+        self.user_profile = user_profile
+        self.marshal = lambda obj: obj
+        self.demarshal = lambda obj: obj
+
+    @contextmanager
+    def __call__(self, marshal=None, demarshal=None):
+        # type: (Optional[Callable[[Any], Text]], Optional[Callable[[Any], Text]]) -> Iterator[StateHandler]
+        prev_marshal = self.marshal
+        prev_demarshal = self.demarshal
+        self.marshal = marshal or self.marshal
+        self.demarshal = demarshal or self.demarshal
+        yield self
+        self.marshal = prev_marshal
+        self.demarshal = prev_demarshal
+
+    def __getitem__(self, key):
+        # type: (Text) -> Text
+        return self.demarshal(get_bot_state(self.user_profile, key))
+
+    def __setitem__(self, key, value):
+        # type: (Text, Text) -> None
+        bot_id = self.user_profile.id
+        if bot_id not in StateHandler.state_sizes:
+            StateHandler.state_sizes[bot_id] = get_bot_state_size(self.user_profile)
+        old_entry_size = get_bot_state_size(self.user_profile, key)
+        new_entry_size = len(key) + len(value)
+        old_state_size = StateHandler.state_sizes[bot_id]
+        new_state_size = old_state_size + (new_entry_size - old_entry_size)
+        if new_state_size > self.state_size_limit:
+            raise RuntimeError("Cannot set state. Request would require {} bytes storage."
+                               "The current storage limit is {}".format(new_state_size, self.state_size_limit))
```

Comment:
I think we want to define a special StateHandlerError exception class to use for the exceptions in this subsystem.